components/layout/layout_impl.rs

```diff
@@ -953,6 +958,45 @@ impl LayoutThread {
         self.stylist
             .force_stylesheet_origins_dirty(sheet_origins_affected_by_device_change);
     }
+
+    /// Main entry point for queries that require [PostCompositeQueryHelper].
+    fn process_post_composite_query<T, Node>(
+        &self,
+        node: Node,
+        process_query_internals: fn(node: Node, helper: PostCompositeQueryHelper) -> T,
+    ) -> T {
+        let scroll_offset_getter = &|fragment: &BoxFragment| {
+            let scroll_id = fragment
+                .base
+                .tag
+                .map(|tag| ExternalScrollId(tag.to_display_list_fragment_id(), self.id.into()));
+            scroll_id
+                .and_then(|id| self.scroll_offsets.borrow().get(&id).cloned())
+                .map(|offset| {
+                    UntypedVector2D::new(Au::from_f32_px(offset.x), Au::from_f32_px(offset.y))
+                })
+                .unwrap_or_default()
+        };
+
+        let root_scroll_offset_getter = &|| {
```

Comment:
What does "root scroll offset" refer to?

Scroll offset of the root element in viewport?

Reply:
Yes, it's specifically referring to the root scroll node which represent the viewport. Added a documentation  in `PostCompositeHelper` struct, thanks!